Objective-C/CBLProperties.mm

```diff
@@ -146,97 +145,403 @@ - (BOOL) containsObjectForKey: (NSString*)key {
+            subdoc = [self subdocumentWithRoot: nullptr withProperties: value forKey: key];
+        return subdoc;
+    } else if ([value isKindOfClass: [NSArray class]]) {
+        NSArray* oldArray = $castIf(NSArray, oldValue);
+        NSArray* nuArray = (NSArray*)value;
+        NSMutableArray* result = [NSMutableArray arrayWithCapacity: nuArray.count];
+        for (int i = 0; i < nuArray.count; i++) {
+            id oldValue = oldArray && i < oldArray.count ? oldArray[i] : nil;
+            id nuValue = nuArray[i];
+            result[i] = [self convertValue: nuValue oldValue: oldValue forKey: key];
+        }
+        
+        // Invalidate:
+        if (oldArray && oldArray.count > nuArray.count) {
+            for (int i = (int)nuArray.count; i < oldArray.count; i++) {
+                CBLSubdocument* oldSubdoc = $castIf(CBLSubdocument, oldArray[i]);
+                [oldSubdoc invalidate];
+            }
+        } else {
+            CBLSubdocument* oldSubdoc = $castIf(CBLSubdocument, oldValue);
+            [oldSubdoc invalidate];
+        }
+        return result;
+    } else {
+        CBLSubdocument* nuSubdoc = $castIf(CBLSubdocument, value);
+        [self attachSubdocument: nuSubdoc forKey: key];
+        
+        if ([oldValue isKindOfClass: [CBLSubdocument class]])
+            [oldValue invalidate];
+    }
+    return value;
+}
+
+
+- (NSDictionary*) currentPropertiesAsJson: (BOOL)asJson {
+    NSMutableDictionary* result;
+    NSMutableSet *changedNames = [_changedNames mutableCopy];
+    if (_root) {
+        int count = FLDict_Count(_root);
+        result = [NSMutableDictionary dictionaryWithCapacity: count];
+        if (count > 0) {
+            FLDictIterator iter;
+            FLDictIterator_Begin(_root, &iter);
+            do {
+                NSString* key = [self fleeceValueToKeyString: FLDictIterator_GetKey(&iter)];
+                id value = _properties[key];
+                if (value) {
+                    if (asJson)
+                        value = [self jsonValueForValue: value];
+                    [changedNames removeObject: key];
+                } else {
+                    value = [self fleeceValueToObject: FLDictIterator_GetValue(&iter)
+                                               forKey: key asJson: asJson];
+                    [self cacheValue: value forKey: key changed: NO];
+                }
+                result[key] = value;
+            } while (FLDictIterator_Next(&iter));
+        }
+        return result;
+    }
+    if (changedNames.count > 0) {
+        if (!result)
+            result = [NSMutableDictionary dictionaryWithCapacity: changedNames.count];
+        for (NSString* key in _changedNames) {
+            id value = _properties[key];
+            result[key] = asJson ? [self jsonValueForValue: value] : value;
+        }
+    }
+    return result;
+}
+
+
+- (id) jsonValueForValue: (id)value {
+    if ([value conformsToProtocol: @protocol(CBLJSONEncoding)]) {
```

Comment:
In most case, when setting a NSDate object, it will get converted to string. BUT there is one case that the NSDate objects will not get converted are when they are in an array.